Panic with nil HTTP rule

Passing an ingress with no HTTP block causes a panic here https://github.com/kubernetes/ingress/blob/master/core/pkg/ingress/controller/controller.go#L630

The HTTP field is nilable:  `HTTP *HTTPIngressRuleValue`

e.g. 

```
spec:
  rules:
  - host: logs.example.com
  tls:
...
```

```
E0202 04:40:03.762472       5 runtime.go:64] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
/usr/local/google/home/beeps/goproj/src/k8s.io/ingress/vendor/k8s.io/kubernetes/pkg/util/runtime/runtime.go:70
/usr/local/google/home/beeps/goproj/src/k8s.io/ingress/vendor/k8s.io/kubernetes/pkg/util/runtime/runtime.go:63
/usr/local/google/home/beeps/goproj/src/k8s.io/ingress/vendor/k8s.io/kubernetes/pkg/util/runtime/runtime.go:49
/usr/lib/google-golang/src/runtime/asm_amd64.s:514
/usr/lib/google-golang/src/runtime/panic.go:489
/usr/lib/google-golang/src/runtime/panic.go:63
/usr/lib/google-golang/src/runtime/signal_unix.go:290
/usr/local/google/home/beeps/goproj/src/k8s.io/ingress/core/pkg/ingress/controller/controller.go:630
/usr/local/google/home/beeps/goproj/src/k8s.io/ingress/core/pkg/ingress/controller/controller.go:371
/usr/local/google/home/beeps/goproj/src/k8s.io/ingress/core/pkg/ingress/controller/controller.go:157
/usr/local/google/home/beeps/goproj/src/k8s.io/ingress/core/pkg/task/queue.go:86
/usr/local/google/home/beeps/goproj/src/k8s.io/ingress/core/pkg/task/queue.go:49
/usr/local/google/home/beeps/goproj/src/k8s.io/ingress/vendor/k8s.io/kubernetes/pkg/util/wait/wait.go:96
/usr/local/google/home/beeps/goproj/src/k8s.io/ingress/vendor/k8s.io/kubernetes/pkg/util/wait/wait.go:97
/usr/local/google/home/beeps/goproj/src/k8s.io/ingress/vendor/k8s.io/kubernetes/pkg/util/wait/wait.go:52
/usr/local/google/home/beeps/goproj/src/k8s.io/ingress/core/pkg/task/queue.go:49
/usr/lib/google-golang/src/runtime/asm_amd64.s:2197
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x131fb54]

goroutine 56 [running]:
k8s.io/ingress/vendor/k8s.io/kubernetes/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
        /usr/local/google/home/beeps/goproj/src/k8s.io/ingress/vendor/k8s.io/kubernetes/pkg/util/runtime/runtime.go:56 +0x126
panic(0x1496800, 0x1ec93a0)
        /usr/lib/google-golang/src/runtime/panic.go:489 +0x2e1
k8s.io/ingress/core/pkg/ingress/controller.(*GenericController).getBackendServers(0xc4200da0e0, 0x1, 0x2, 0xc4204f3d7c, 0x2, 0x2, 0xc4204f3d78)
        /usr/local/google/home/beeps/goproj/src/k8s.io/ingress/core/pkg/ingress/controller/controller.go:630 +0x2f4
k8s.io/ingress/core/pkg/ingress/controller.(*GenericController).sync(0xc4200da0e0, 0x140c300, 0xc42026fe10, 0xc42026fe10, 0xc42026fd00)
        /usr/local/google/home/beeps/goproj/src/k8s.io/ingress/core/pkg/ingress/controller/controller.go:371 +0xa6
k8s.io/ingress/core/pkg/ingress/controller.(*GenericController).(k8s.io/ingress/core/pkg/ingress/controller.sync)-fm(0x140c300, 0xc42026fe10, 0xa, 0xc4204f3e00)
        /usr/local/google/home/beeps/goproj/src/k8s.io/ingress/core/pkg/ingress/controller/controller.go:157 +0x3e
k8s.io/ingress/core/pkg/task.(*Queue).worker(0xc42012a480)
        /usr/local/google/home/beeps/goproj/src/k8s.io/ingress/core/pkg/task/queue.go:86 +0xef
k8s.io/ingress/core/pkg/task.(*Queue).(k8s.io/ingress/core/pkg/task.worker)-fm()
        /usr/local/google/home/beeps/goproj/src/k8s.io/ingress/core/pkg/task/queue.go:49 +0x2a
k8s.io/ingress/vendor/k8s.io/kubernetes/pkg/util/wait.JitterUntil.func1(0xc4205cdfa8)
        /usr/local/google/home/beeps/goproj/src/k8s.io/ingress/vendor/k8s.io/kubernetes/pkg/util/wait/wait.go:96 +0x5e
k8s.io/ingress/vendor/k8s.io/kubernetes/pkg/util/wait.JitterUntil(0xc420729fa8, 0x12a05f200, 0x0, 0x1, 0xc420052720)
        /usr/local/google/home/beeps/goproj/src/k8s.io/ingress/vendor/k8s.io/kubernetes/pkg/util/wait/wait.go:97 +0xad
k8s.io/ingress/vendor/k8s.io/kubernetes/pkg/util/wait.Until(0xc4205cdfa8, 0x12a05f200, 0xc420052720)
        /usr/local/google/home/beeps/goproj/src/k8s.io/ingress/vendor/k8s.io/kubernetes/pkg/util/wait/wait.go:52 +0x4d
k8s.io/ingress/core/pkg/task.(*Queue).Run(0xc42012a480, 0x12a05f200, 0xc420052720)
        /usr/local/google/home/beeps/goproj/src/k8s.io/ingress/core/pkg/task/queue.go:49 +0x5e
created by k8s.io/ingress/core/pkg/ingress/controller.GenericController.Start
        /usr/local/google/home/beeps/goproj/src/k8s.io/ingress/core/pkg/ingress/controller/controller.go:1047 +0x1c2
```

--- a/core/pkg/ingress/controller/annotations_test.go
+++ b/core/pkg/ingress/controller/annotations_test.go
@@ -28,10 +28,10 @@ import (
 )
 
 const (
-	annotation_secureUpstream = "ingress.kubernetes.io/secure-backends"
-	annotation_upsMaxFails    = "ingress.kubernetes.io/upstream-max-fails"
-	annotation_upsFailTimeout = "ingress.kubernetes.io/upstream-fail-timeout"
-	annotation_passthrough    = "ingress.kubernetes.io/ssl-passthrough"
+	annotationSecureUpstream = "ingress.kubernetes.io/secure-backends"
+	annotationUpsMaxFails    = "ingress.kubernetes.io/upstream-max-fails"
+	annotationUpsFailTimeout = "ingress.kubernetes.io/upstream-fail-timeout"
+	annotationPassthrough    = "ingress.kubernetes.io/ssl-passthrough"
 )
 
 type mockCfg struct {
@@ -106,9 +106,9 @@ func TestSecureUpstream(t *testing.T) {
 		annotations map[string]string
 		er          bool
 	}{
-		{map[string]string{annotation_secureUpstream: "true"}, true},
-		{map[string]string{annotation_secureUpstream: "false"}, false},
-		{map[string]string{annotation_secureUpstream + "_no": "true"}, false},
+		{map[string]string{annotationSecureUpstream: "true"}, true},
+		{map[string]string{annotationSecureUpstream: "false"}, false},
+		{map[string]string{annotationSecureUpstream + "_no": "true"}, false},
 		{map[string]string{}, false},
 		{nil, false},
 	}
@@ -131,9 +131,9 @@ func TestHealthCheck(t *testing.T) {
 		eumf        int
 		euft        int
 	}{
-		{map[string]string{annotation_upsMaxFails: "3", annotation_upsFailTimeout: "10"}, 3, 10},
-		{map[string]string{annotation_upsMaxFails: "3"}, 3, 0},
-		{map[string]string{annotation_upsFailTimeout: "10"}, 0, 10},
+		{map[string]string{annotationUpsMaxFails: "3", annotationUpsFailTimeout: "10"}, 3, 10},
+		{map[string]string{annotationUpsMaxFails: "3"}, 3, 0},
+		{map[string]string{annotationUpsFailTimeout: "10"}, 0, 10},
 		{map[string]string{}, 0, 0},
 		{nil, 0, 0},
 	}
@@ -164,9 +164,9 @@ func TestSSLPassthrough(t *testing.T) {
 		annotations map[string]string
 		er          bool
 	}{
-		{map[string]string{annotation_passthrough: "true"}, true},
-		{map[string]string{annotation_passthrough: "false"}, false},
-		{map[string]string{annotation_passthrough + "_no": "true"}, false},
+		{map[string]string{annotationPassthrough: "true"}, true},
+		{map[string]string{annotationPassthrough: "false"}, false},
+		{map[string]string{annotationPassthrough + "_no": "true"}, false},
 		{map[string]string{}, false},
 		{nil, false},
 	}

--- a/core/pkg/ingress/controller/controller.go
+++ b/core/pkg/ingress/controller/controller.go
@@ -620,9 +620,8 @@ func (ic *GenericController) getBackendServers() ([]*ingress.Backend, []*ingress
 			}
 
 			if rule.HTTP == nil &&
-				len(ing.Spec.TLS) == 0 &&
 				host != defServerName {
-				glog.V(3).Infof("ingress rule %v/%v does not contains HTTP or TLS rules. using default backend", ing.Namespace, ing.Name)
+				glog.V(3).Infof("ingress rule %v/%v does not contains HTTP rules. using default backend", ing.Namespace, ing.Name)
 				server.Locations[0].Backend = defBackend.Name
 				continue
 			}

